Citra-qt: Add multiplayer ui

jroweboy opened the pull request:
## Design Goals

Multiplayer works by emulating local wireless connections and serving the packets to others across the internet. Since most games aren't designed to support large groups of people in the same location, the goal of this design is to focus on smaller rooms of people that are playing a single game instead of larger rooms where people play whatever game they feel like. The design is meant to encourage the idea of "I want to play X game with others" and to help guide them to others that also want to play that game.

With this in mind, the room is limited to 16 people in order to encourage many smaller rooms instead of a few larger rooms. The lobby also allows users to filter by games that I own, hide full rooms from the lobby, and other filtering options. Inside the room, there is a chat system which is only meant to be simple enough to provide bare bones communication between players (useful for things like helping users determine game mode to play and other coordination tasks)

Due to limitations on time, I decided that it would be best to focus on the use case listed above, of random people playing together. Its easier to get a few dedicated rooms set up that people can join, rather than having people host their own rooms and having them deal with port forwarding and so on. There wasn't enough time to add support for NAT traversal for the first release, so prioritizing players using dedicated rooms through the lobby was the plan.

In summary, the primary goal of the first release of the UI is meant to help someone find a group to play a specific game with. From there, secondary goals are to let people host their own rooms, direct connect, and so on.

## Screenshots

Lobby: 
![lobby](https://user-images.githubusercontent.com/952515/32458960-6356951e-c2eb-11e7-9b3b-33342f7ef756.png)

Features of lobby:
 * Easy to change the nickname you want to connect the room with.
 * Filter rooms by search (searches on room name, preferred game and host name), games I own (by game id so it doesn't handle region differences currently), and Hide full games (max players)
 * Sortable columns
 * Doesn't hammer citra services by automatically refreshing (intended)

Rooms:
![connected as client](https://user-images.githubusercontent.com/952515/32459824-1e636d08-c2ee-11e7-96d4-0ba053a258b8.png)

Features of Room:
  * Simple. Has only player listing and chat.
  * Disconnect is different from close window.

Create Room:
![create room](https://user-images.githubusercontent.com/952515/32460061-b9ed6940-c2ee-11e7-9c32-eb8c9a152678.png)

Features of create room:
 * Simple. Designed to put the "Player list and Chat" below it once the room is created, but that had to be scrapped from this release due to technical issues with updating options after the room hosting is started.
 * Fields like nickname and room name are stored in uisettings.

Direct Connect:
![direct connect](https://user-images.githubusercontent.com/952515/32460228-58ce05b0-c2ef-11e7-87eb-51f3551bd29d.png)

Features of direct connect:
 * Ready to support NAT traversal (TODO: remove text about traversal server)
 * Fields are stored in uisettings.

## Code Design

**TODO** (sorry ran outta time writing this again!)

## TODO

- [x] Remove text about traversal server in direct connect (since that didn't make the first release)
- [x] Add "Connected" text to the status bar next to the connected icon (along with a tool tip telling people that they can double click it to open the room)
- [x] Fix `QObject::connect: signal not found in Lobby` error
- [ ] ~~Make the multiplayer windows not parented so you can minimize them independently~~ Won't do

## Follow up changes

- [ ] NAT traversal support to make self hosting more useful
- [ ] Integrated NAT feedback to help people diagnose if they can host
- [x] ~~make citra faster~~

Also major apologies for the single commit. I will split it up if reviewing file by file is proving to be too much

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/3073)
<!-- Reviewable:end -->

BynariStar commented:
As discussed on Discord:

- After closing the multiplayer window (but not disconnecting) and reopening the lobby, it prompts disconnection while trying to reopen the room you're currently still in. Should probably change this to just open the room.
- Should probably change the "Chat" button to something more along the lines of "Current Room" to avoid confusion
- The connection status icon IS TOO DAMN SMALL (scaling issue with HiDPI screens)

esQmo commented:
Please add a back button to the chat client beacause when you select a room and go in, there is no way to go back to the room browser, you have to close the chat client windows and then re-open it again from the Multiplayer tab. to be able to, say, select another rom. Or you can make it open in a new window instead of that same.

BreadFish64 commented on src/citra_qt/main.cpp:
```diff
@@ -714,6 +758,27 @@ void GMainWindow::OnMenuRecentFile() {
+    announce_multiplayer_session->Stop();
+    QMessageBox::warning(
+        this, tr("Error"),
+        tr("Announcing the room failed. \nThe room will not get listed publicy. \nError: ") +
```
listed publicly

baka0815 replied:
Also the spaces before the \n are not really necessary.

baka0815 commented on dist/icons/license.md:
```diff
@@ -0,0 +1,7 @@
+connect.png       CC BY-ND 3.0 (aquired from https://icons8.com)
+disconnect.png    CC BY-ND 3.0 (aquired from https://icons8.com)
+failed.png        Free for non-commercial use
+lock.png         CC BY-ND 3.0 (aquired from https://icons8.com)
```
Indentation

baka0815 commented on src/citra_qt/main.ui:
```diff
@@ -206,6 +221,43 @@
+   </action>
+  <action name="action_Connect_To_Room">
+   <property name="text">
+    <string>Direct Connect to Room</string>
```
What's the "Direct" here for? Isn't "Connect to Room" enough?

Is the capitalization correct the way it is? Shouldn't "Room" be lower case?
So "Direct[ly] connect to room" or "Connect to room"?
The same comment applies to "Create Room" ("Create room") and "Browse Public Game Lobby" ("Browse public game lobby").

jroweboy replied:
The current wording is consistent with the existing menu items. All of the menu items capitalize the first letter of each word.

baka0815 commented on src/citra_qt/main.cpp:
```diff
@@ -714,6 +758,27 @@ void GMainWindow::OnMenuRecentFile() {
+        return;
+    }
+    network_status->setPixmap(QPixmap(":/icons/disconnected.png"));
+    ui.action_Chat->setDisabled(true);
```
I'm not into Qt enough to know, but you're using "setDisabled(true)" here and "setEnabled(false)" in another place.
Shouldn't that all be using "setEnabled(true)" or "setEnabled(false)" respectively?

At least I find it confusing.

baka0815 commented on src/citra_qt/multiplayer/lobby.cpp:
```diff
@@ -0,0 +1,314 @@
+const QString Lobby::PasswordPrompt() {
+    bool ok;
+    const QString text =
+        QInputDialog::getText(this, tr("Password Required to Join"), tr("Password:"),
```
Password **r**equired to **j**oin

jroweboy replied:
This is the title of the window so the capitalization is correct.

baka0815 commented on src/citra_qt/multiplayer/lobby.h:
```diff
@@ -0,0 +1,118 @@
+     * Pulls the list of rooms from network and fills out the lobby model with the results
+     */
+    void OnRefreshLobby();
+    /**
```
Missing newline?

baka0815 commented on src/citra_qt/multiplayer/lobby.h:
```diff
@@ -0,0 +1,118 @@
+    void ResetModel();
+
+    /**
+     * Prompts for a password. Returns an empty QString in the user either did not provide a
```
QString **if** the user

baka0815 commented on src/citra_qt/multiplayer/validation.h:
```diff
@@ -0,0 +1,28 @@
+
+namespace Validation {
+/// room name can be alphanumeric and " " "_" "." and "-"
+static const QRegExp room_name_regex("^[a-zA-Z0-9._- ]+$");
```
Is "." correct here? Shouldn't it be "\\."?

lioncash replied:
static can be omitted from all of these. `const` internally links by default.

baka0815 commented on src/citra_qt/multiplayer/validation.h:
```diff
@@ -0,0 +1,28 @@
+static const QValidator* room_name = new QRegExpValidator(room_name_regex);
+
+/// nickname can be alphanumeric and " " "_" "." and "-"
+static const QRegExp nickname_regex("^[a-zA-Z0-9._- ]+$");
```
Same as room name.

jroweboy replied:
Thanks for the review so far, but I would love it if you batched the comments into a single review so i don't get flooded with emails about it :)

(Also, yes the regex is correct. Its a character class, so you don't need to escape the period https://www.regular-expressions.info/refcharclass.html)

baka0815 replied:
Sorry about the noise, I'll keep that in mind for the next time.

MerryMage commented on src/citra_qt/main.h:
```diff
@@ -48,6 +62,9 @@ class GMainWindow : public QMainWindow {
     void UpdateUITheme();
+    void ChangeRoomState();
+
+    GameList* game_list;
```
Duplicate `game_list`.

MerryMage commented:
RE CI: Ignore continuous-integration/travis-ci/push and continuous-integration/appveyor/branch, that was my error.

B3n30 commented:
Reviewed 8 of 35 files at r1, 1 of 3 files at r2, 8 of 8 files at r3.
Review status: 17 of 35 files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---

*[src/citra_qt/multiplayer/chat_room.cpp, line 154 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3YF3bKARxMRT7XW5io:-L3YF3bKARxMRT7XW5ip:by27tax) ([raw file](https://github.com/citra-emu/citra/blob/9c482ee1fa9e373d166b00fede5ad13bf5402827/src/citra_qt/multiplayer/chat_room.cpp#L154)):*
> ```cpp
>             return;
>         }
>         auto player = std::distance(members.begin(), it);
> ```

That way the color of the chat messages will change as soon as a player disconnects

---

*[src/citra_qt/multiplayer/chat_room.cpp, line 178 at r3](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3YFHd2i33dShHsYfdV:-L3YFHd2i33dShHsYfdW:b2ss7xm) ([raw file](https://github.com/citra-emu/citra/blob/9c482ee1fa9e373d166b00fede5ad13bf5402827/src/citra_qt/multiplayer/chat_room.cpp#L178)):*
> ```cpp
>                 LOG_INFO(Network, "Chat message received from unknown player");
>             }
>             auto player = std::distance(members.begin(), it);
> ```

see comment above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3073)*
<!-- Sent from Reviewable.io -->

B3n30 commented:
Reviewed 26 of 35 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks broke.

---

*[src/citra_qt/multiplayer/client_room.cpp, line 13 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e0MwbFbKhP34KMmJV:-L3e0MwbFbKhP34KMmJW:bqx6ynr) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/multiplayer/client_room.cpp#L13)):*
> ```cpp
> #include <QTime>
> #include <QtConcurrent/QtConcurrentRun>
> ```

remove blank line

---

*[src/citra_qt/multiplayer/direct_connect.ui, line 85 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e2BgNP4pXYv5EYGpt:-L3e2BgNP4pXYv5EYGpu:bvss8fd) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/multiplayer/direct_connect.ui#L85)):*
> ```XML
>              <widget class="QLineEdit" name="ip">
>               <property name="toolTip">
>                <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;IPv4 address of the host&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
> ```

Is it necessary to put these strings into that ugly html code?

---

*[src/citra_qt/multiplayer/host_room.cpp, line 51 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e2khcTR31NPnQRht9:-L3e2khcTR31NPnQRhtA:bs8pe6y) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/multiplayer/host_room.cpp#L51)):*
> ```cpp
>     ui->port->setText(UISettings::values.room_port);
>     ui->max_player->setValue(UISettings::values.max_player);
>     int index = ui->host_type->findData(UISettings::values.host_type);
> ```

host type isn't restored properly. Not sure why though

---

*[src/citra_qt/multiplayer/lobby.cpp, line 175 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e3YyY8Od_9YZyLDKm:-L3e3YyY8Od_9YZyLDKn:bmdshz2) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/multiplayer/lobby.cpp#L175)):*
> ```cpp
>         ui->refresh_list->setText(tr("Refreshing"));
>     } else {
>         // TODO(jroweboy): Display an error box about network couldn't be started
> ```

about announce couldn't get started.

---

*[src/citra_qt/util/clickable_label.cpp, line 1 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e4GqaWDXIr2Y6MJe_:-L3e4GqaWDXIr2Y6MJea:b-nzk1ak) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/util/clickable_label.cpp#L1)):*
> ```cpp
> #include "util/clickable_label.h"
> ```

Copyright header?

---

*[src/citra_qt/util/clickable_label.h, line 1 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3073#-L3e4L-yCMGykAYrL1QL:-L3e4L-yCMGykAYrL1QM:bxo2i1) ([raw file](https://github.com/citra-emu/citra/blob/bbf148ffde8b29c2b265c2e2c17c85d297841f44/src/citra_qt/util/clickable_label.h#L1)):*
> ```objc
> #ifndef CLICKABLELABEL_H
> ```

#pragma once
Copyright header

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3073)*
<!-- Sent from Reviewable.io -->

adityaruplaha commented:
When will this be merged into nightly? Its been in canary for quite some time.

BreadFish64 commented:
@adityalruplaha this is the last piece of the puzzle for multiplayer, all the other parts of multiplayer need to merged before this

CaptV0rt3x commented:
@jroweboy  Please check out #3432 . According to @B3n30 , it is caused by this PR.

FearlessTobi commented:
Just noticed that you are able to chose a name with spaces at the end. They should be trimmed away imo.

jroweboy commented:
A long time ago, I had pestered neobrain for some feedback about the UI that I planned to apply in some shape or form to this before merging. Here was the conversation from the logs so i can reference later: https://pastebin.com/d6y3WB2f

baka0815 commented on dist/icons/license.md:
```diff
@@ -0,0 +1,7 @@
+The icons in this folder have the following licenses:
+checked.png       Free for non-commercial use
+connect.png       CC BY-ND 3.0 (aquired from https://icons8.com)
```
The filename is "connected.png"

baka0815 commented on dist/icons/license.md:
```diff
@@ -0,0 +1,7 @@
+The icons in this folder have the following licenses:
+checked.png       Free for non-commercial use
+connect.png       CC BY-ND 3.0 (aquired from https://icons8.com)
+disconnect.png    CC BY-ND 3.0 (aquired from https://icons8.com)
```
The filename is "disconnected.png"

chris062689 commented:
What are we planning on doing from a UI perspective with multiple netVersions?

We will likely run into this scenario when this gets merged into nightly, and we push out a new net version for Canary builds. 

Should we grey out servers that run a different net version? 
What should we do from a UI perspective?

chris062689 commented:
The version mismatch error is a bit cryptic. 
`you are using a different version of Citra-Local-Wifi???`

Perhaps a better dialog message would be
`ERROR: VERSION_MISMATCH You are attempting to connect to a room running a different multiplayer version. Please ensure that you are running the latest version of Citra.`

adityaruplaha commented:
Yes. That would also discourage unofficial builds, as they will most likely
be outdated very fast.

Adityarup Laha

On 08-Mar-2018 7:40 pm, "Flame Sage" <notifications@github.com> wrote:

> The version mismatch error is a bit cryptic.
> you are using a different version of Citra-Local-Wifi???
>
> Perhaps a better dialog message would be
> ERROR: VERSION_MISMATCH You are attempting to connect to a room running a
> different multiplayer version. Please ensure that you are running the
> latest version of Citra.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/citra-emu/citra/pull/3073#issuecomment-371497250>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AdRkQ_XeG42YvUTKhzDlGvadA4GSqckrks5tcTvvgaJpZM4QSIv6>
> .
>

Schplee commented:
*Recommended by Flame Sage to post here*

* Perhaps the lobby should default sort by player count, so that users can have an easier time seeing which lobbies are populated. With how many lobbies exist now, it can become a hassle to sort without pressing a button.

* I'm not sure if it's already planned, but currently the tabs' size can't be changed: https://i.imgur.com/m1Raj4p.png

jroweboy commented:
@Schplee Don't really like that as the default sort. Sort by name seems so much more intuitive as you want to try to find lobbies that are close to you so you have the best latency usually.

wasn't planning on making the tabs resizable, but i can do that

also note to self: add lobby host username in the list of things that can be filtered by. no reason not to add that imo

jroweboy commented:
Conflict fixed. Added a little "Connected" status bar message as well as migrated to using the new themed icons so if someone wants to make me some dark themed versions for the connection/disconnected icons, go ahead, and we can stick them in there. I plan to fix up the UI state stuff soon (TM) so we can merge it so others feel more comfortable fixing issues with this...

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+    }
+
+private:
+    ChatMessage() {}
```
This can likely just be removed if the intention is to not allow it to be used, since another constructor is already provided.

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+
+private:
+    ChatMessage() {}
+    const QList<QString> player_color = {
```
This should probably be `static const`. It can likely also just be a std::array of const char*'s. There's no real need to allocate it all on the heap in Qt strings, when it's just being printed out.

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+public:
+    explicit ChatMessage(const Network::ChatEntry& chat, QTime ts = {}) {
+        /// Convert the time to their default locale defined format
+        static QLocale locale;
```
Is there a reason this just doesn't construct the default instance on each invocation? Is it really that performance intensive?

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+    }
+
+private:
+    const QString system_color = "#888888";
```
This can be: `static constexpr const char system_color[] = "#888888"`

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+    QString message;
+};
+
+ChatRoom::ChatRoom(QWidget* parent) : ui(new Ui::ChatRoom) {
```
make_unique

jroweboy replied:
We don't use unique ptrs for UI files anywhere in the application. I'm not sure if we *should* but I'm just following the practice that the rest of the code uses.

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+
+void ChatRoom::OnSendChat() {
+    if (auto room = Network::GetRoomMember().lock()) {
+        if (room->GetState() == Network::RoomMember::State::Joined) {
```
This if statement can be an early return, unindenting all the code within it.

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+    // TODO(B3N30): Remember which row is selected
+    player_list->removeRows(0, player_list->rowCount());
+    for (const auto& member : member_list) {
+        if (member.nickname == "")
```
`if (member.nickname.empty())`

lioncash commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+            continue;
+        QList<QStandardItem*> l;
+        std::vector<std::string> elements = {member.nickname, member.game_info.name};
+        for (auto& item : elements) {
```
`const auto&`

lioncash commented on src/citra_qt/multiplayer/client_room.cpp:
```diff
@@ -0,0 +1,121 @@
+
+ClientRoomWindow::ClientRoomWindow(QWidget* parent)
+    : QDialog(parent, Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::WindowSystemMenuHint),
+      ui(new Ui::ClientRoom) {
```
make_unique

jroweboy replied:
in setupUi, qt parents all of the ui elements to the parent widget, so it should be the one responsible for managing the memory as I understand it.

jroweboy replied:
@wwylele if thats the case, then we should make this change for the other places that have the issue.

wwylele replied:
@jroweboy AFAIK most if not all existing UI in our code base is using `std::unique_ptr`

jroweboy replied:
Nope

* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/configuration/configure_debug.cpp#L9
* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/configuration/configure_dialog.cpp#L10
* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/configuration/configure_general.cpp#L13
* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/configuration/configure_graphics.cpp#L11
* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/configuration/configure_system.cpp#L28
* https://github.com/citra-emu/citra/blob/2d372bae1409cd8088c9c6b03fb7334a1e4357f0/src/citra_qt/aboutdialog.cpp#L12 (This one deletes it appropriately though)

wwylele replied:
@jroweboy for all examples you gave they are declared as `unique_ptr` in the header file, although constructed by `new`, which is, though not preferable, still fine. However yours are just simply declared as raw pointer.

jroweboy replied:
wasn't aware that you can pass a raw pointer to the constructor of unique ptr like that. (sorry for not checking the header) will fix, thanks.

lioncash commented on src/citra_qt/multiplayer/chat_room.h:
```diff
@@ -0,0 +1,57 @@
+    void AppendChatMessage(const QString&);
+    bool ValidateMessage(const std::string&);
+    QStandardItemModel* player_list;
+    Ui::ChatRoom* ui;
```
This should be a unique_ptr, otherwise you leak resources.

lioncash commented on src/citra_qt/multiplayer/client_room.h:
```diff
@@ -0,0 +1,38 @@
+    void UpdateView();
+
+    QStandardItemModel* player_list;
+    Ui::ClientRoom* ui;
```
This should be a unique_ptr otherwise you leak resources.

lioncash commented on src/citra_qt/multiplayer/direct_connect.cpp:
```diff
@@ -0,0 +1,132 @@
+
+DirectConnectWindow::DirectConnectWindow(QWidget* parent)
+    : QDialog(parent, Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::WindowSystemMenuHint),
+      ui(new Ui::DirectConnect) {
```
make_unique

lioncash commented on src/citra_qt/multiplayer/direct_connect.cpp:
```diff
@@ -0,0 +1,132 @@
+#include "network/network.h"
+#include "ui_direct_connect.h"
+
+enum class ConnectionType : u8 { TRAVERSAL_SERVER, IP };
```
enum class formatting would be `TraversalServer, IP`

lioncash commented on src/citra_qt/multiplayer/direct_connect.cpp:
```diff
@@ -0,0 +1,132 @@
+        ShowError(NetworkMessage::USERNAME_NOT_VALID);
+    }
+    if (const auto member = Network::GetRoomMember().lock()) {
+        if (member->IsConnected()) {
```
```cpp
if (member->IsConnected() && !NetworkMessage::WarnDisconnect()) {
    return;
}
```

lioncash commented on src/citra_qt/multiplayer/direct_connect.cpp:
```diff
@@ -0,0 +1,132 @@
+    bool isValid = true;
+    if (!ui->nickname->hasAcceptableInput()) {
+        isValid = false;
+        ShowError(NetworkMessage::USERNAME_NOT_VALID);
```
Should this not just return here rather than waiting until it hits the check further below?

jroweboy replied:
The original idea was to show form errors inline, and an early return doesn't let all validators run. Currently its just a little more annoying as you get multiple popups (one for each error), so i'm fine changing it.

lioncash commented on src/citra_qt/multiplayer/direct_connect.h:
```diff
@@ -0,0 +1,39 @@
+    void EndConnecting();
+
+    QFutureWatcher<void>* watcher;
+    Ui::DirectConnect* ui;
```
This should be a unique_ptr otherwise you leak resources

lioncash commented on src/citra_qt/multiplayer/host_room.cpp:
```diff
@@ -0,0 +1,170 @@
+HostRoomWindow::HostRoomWindow(QWidget* parent, QStandardItemModel* list,
+                               std::shared_ptr<Core::AnnounceMultiplayerSession> session)
+    : QDialog(parent, Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::WindowSystemMenuHint),
+      ui(new Ui::HostRoom), announce_multiplayer_session(session), game_list(list) {
```
make_unique

lioncash commented on src/citra_qt/multiplayer/host_room.h:
```diff
@@ -0,0 +1,73 @@
+    std::weak_ptr<Core::AnnounceMultiplayerSession> announce_multiplayer_session;
+    QStandardItemModel* game_list;
+    ComboBoxProxyModel* proxy;
+    Ui::HostRoom* ui;
```
This should be a unique_ptr otherwise you leak resources.

lioncash commented on src/citra_qt/util/clickable_label.h:
```diff
@@ -0,0 +1,23 @@
+    Q_OBJECT
+
+public:
+    explicit ClickableLabel(QWidget* parent = Q_NULLPTR, Qt::WindowFlags f = Qt::WindowFlags());
```
= nullptr

lioncash commented on src/citra_qt/util/clickable_label.h:
```diff
@@ -0,0 +1,23 @@
+
+#include <QLabel>
+#include <QWidget>
+#include <Qt>
```
I would remove including the main Qt header in another header file.

lioncash commented on src/citra_qt/util/clickable_label.cpp:
```diff
@@ -0,0 +1,13 @@
+
+ClickableLabel::ClickableLabel(QWidget* parent, Qt::WindowFlags f) : QLabel(parent) {}
+
+ClickableLabel::~ClickableLabel() {}
```
This can be a defaulted destructor

lioncash commented on src/citra_qt/multiplayer/message.h:
```diff
@@ -0,0 +1,53 @@
+class ConnectionError {
+
+public:
+    explicit ConnectionError(const std::string& str) : err(str) {}
```
std::string can be taken by value here and std::moved in the initializer list.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int MemberListRole = Qt::UserRole + 1;
+    static const int MaxPlayerRole = Qt::UserRole + 2;
+
+    LobbyItemMemberList() : LobbyItem() {}
```
This can just be:

```cpp
LobbyItemMemberList() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int MaxPlayerRole = Qt::UserRole + 2;
+
+    LobbyItemMemberList() : LobbyItem() {}
+    LobbyItemMemberList(QList<QVariant> members, u32 max_players) : LobbyItem() {
```
`: LobbyItem()` is redundant.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+
+class LobbyMember {
+public:
+    LobbyMember() {}
```
This can be a defaulted constructor

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+class LobbyMember {
+public:
+    LobbyMember() {}
+    LobbyMember(const LobbyMember& other) {
```
This can just be

```cpp
LobbyMember(const LobbyMember& other) = default;
```

the copy-assignment operator should likely be defaulted as well.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    }
+    LobbyMember(const QString username, u64 title_id, const QString game_name)
+        : username(username), title_id(title_id), game_name(game_name) {}
+    ~LobbyMember() {}
```
This can be a defaulted destructor

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+        title_id = other.title_id;
+        game_name = other.game_name;
+    }
+    LobbyMember(const QString username, u64 title_id, const QString game_name)
```
These can just be non-const and std::moved in the initializer lists.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int HostIPRole = Qt::UserRole + 2;
+    static const int HostPortRole = Qt::UserRole + 3;
+
+    LobbyItemHost() : LobbyItem() {}
```
This can just be:

```cpp
LobbyItemHost() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int HostPortRole = Qt::UserRole + 3;
+
+    LobbyItemHost() : LobbyItem() {}
+    LobbyItemHost(QString username, QString ip, u16 port) : LobbyItem() {
```
```cpp
: LobbyItem()
```
is redundant.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int GameNameRole = Qt::UserRole + 2;
+    static const int GameIconRole = Qt::UserRole + 3;
+
+    LobbyItemGame() : LobbyItem() {}
```
This can just be:

```cpp
LobbyItemGame() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+    static const int GameIconRole = Qt::UserRole + 3;
+
+    LobbyItemGame() : LobbyItem() {}
+    LobbyItemGame(u64 title_id, QString game_name, QPixmap smdh_icon) : LobbyItem() {
```
```cpp
: LobbyItem()
```
is redundant.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+class LobbyItemName : public LobbyItem {
+public:
+    static const int NameRole = Qt::UserRole + 1;
+    LobbyItemName() : LobbyItem() {}
```
This can just be:

```cpp
LobbyItemName() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+public:
+    static const int NameRole = Qt::UserRole + 1;
+    LobbyItemName() : LobbyItem() {}
+    LobbyItemName(QString name) : LobbyItem() {
```
`explicit LobbyItemName`

also

```cpp
: LobbyItem()
```
is redundant.

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+
+class LobbyItem : public QStandardItem {
+public:
+    LobbyItem() : QStandardItem() {}
```
```cpp
LobbyItem() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+class LobbyItem : public QStandardItem {
+public:
+    LobbyItem() : QStandardItem() {}
+    LobbyItem(const QString& string) : QStandardItem(string) {}
```
```cpp
explicit LobbyItem
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+public:
+    LobbyItem() : QStandardItem() {}
+    LobbyItem(const QString& string) : QStandardItem(string) {}
+    virtual ~LobbyItem() override {}
```
```cpp
virtual ~LobbyItem() override = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+class LobbyItemPassword : public LobbyItem {
+public:
+    static const int PasswordRole = Qt::UserRole + 1;
+    LobbyItemPassword() : LobbyItem() {}
```
```cpp
LobbyItemPassword() = default;
```

lioncash commented on src/citra_qt/multiplayer/lobby_p.h:
```diff
@@ -0,0 +1,189 @@
+public:
+    static const int PasswordRole = Qt::UserRole + 1;
+    LobbyItemPassword() : LobbyItem() {}
+    LobbyItemPassword(const bool has_password) : LobbyItem() {
```
```cpp
explicit LobbyItemPassword
```
also `: LobbyItem()` is redundant.

lioncash commented on src/citra_qt/multiplayer/lobby.cpp:
```diff
@@ -0,0 +1,314 @@
+Lobby::Lobby(QWidget* parent, QStandardItemModel* list,
+             std::shared_ptr<Core::AnnounceMultiplayerSession> session)
+    : QDialog(parent, Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::WindowSystemMenuHint),
+      ui(new Ui::Lobby), announce_multiplayer_session(session), game_list(list) {
```
make_unique

lioncash commented on src/citra_qt/multiplayer/lobby.cpp:
```diff
@@ -0,0 +1,314 @@
+    ui->chat->setDisabled(true);
+}
+
+Lobby::~Lobby() {}
```
This can be a defaulted destructor

lioncash commented on src/citra_qt/multiplayer/chat_room.h:
```diff
@@ -0,0 +1,57 @@
+    void ChatReceived(const Network::ChatEntry&);
+
+private:
+    const u32 max_chat_lines = 1000;
```
`static constexpr` or move this into the cpp file.

lioncash commented on src/citra_qt/main.cpp:
```diff
@@ -1184,6 +1333,16 @@ void GMainWindow::closeEvent(QCloseEvent* event) {
 
     render_window->close();
 
+    // Close Multiplayer windows
```
This should probably go within its own function.

lioncash commented on src/citra_qt/main.cpp:
```diff
@@ -1053,6 +1128,80 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
+
+void GMainWindow::OnCloseRoom() {
+    if (auto room = Network::GetRoom().lock()) {
+        if (room->GetState() == Network::Room::State::Open) {
```
These can be inverted to unindent contained code.

jroweboy commented:
Multiplayer UI status:

* Moved almost all the multiplayer handling stuff from main to its own class
* Added a new feature to allow people to see who is playing what in the lobby. (Screenshot below)
* Fixed an issue where you couldn't host a room, and then close it and host another
* Found a new issue where if you try to join a full room, theres no error message. From there, you can't join the next room (you get an error message) and then the next room you try to join makes citra crash :P 
* Haven't address @lioncash 's comments yet. Will be doing that before merge though.
* Blank names seem to be allowed somehow. Need to look more into that

Light theme
![citra-qt_2018-04-05_12-13-16](https://user-images.githubusercontent.com/952515/38383837-c5bb7fee-38ca-11e8-8f5a-4f960077d8b6.png)

Dark theme
![citra-qt_2018-04-05_12-14-08](https://user-images.githubusercontent.com/952515/38383865-dcda88f0-38ca-11e8-83e3-bb675a3c29a6.png)

FearlessTobi commented:
I personally don't think the "Current Players in the room" text is needed. Imo it's fairly self-explanatory what's displayed there.

daniellimws commented:
Why is the first column called "Password"?

baka0815 commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+        }
+        auto player = std::distance(members.begin(), it);
+        ChatMessage m(chat);
+        AppendChatMessage(m.GetPlayerChatMessage(player));
```
Doesn't that change the color of all players if one who joined earlier leaves?

jroweboy replied:
yes. its a hack to add *some* color consistency since theres not a persistent player id assigned by the room and i couldn't convince anyone to add one either.

baka0815 commented on src/citra_qt/multiplayer/chat_room.cpp:
```diff
@@ -0,0 +1,208 @@
+            auto player = std::distance(members.begin(), it);
+            ChatMessage m(chat);
+            room->SendChatMessage(message);
+            AppendChatMessage(m.GetPlayerChatMessage(player));
```
Same as above.
Also can't this duplicated code (``get the id of the player``) be put into it's own function?

tywald commented:
It seems commit 1acbea9 broke a couple of things:
1) The 'File' and 'Emulation' menus are unclickable, doesn't react when you mouseover them. Need to click on 'View', 'Multiplayer' or 'Help' menus then move the mouse towards 'Emulation' and 'File' to access them.

2) Filtering doesn't work in the public room browser.

Tested with Appveyor builds 3b170be2 and eba8953c with a fresh config.

jroweboy commented:
@daniellimws the password column shows an icon if a room has a password on it. the columns are ordered by things that affect whether or not people want to join a room. If someone is just looking to play pokemon, they need to know first if its password protected, as they can't join it if it is. From there, I somewhat arbitrarily chose the ordering of the other columns, but it seemed to be a pretty good ordering.

daniellimws commented:
@jroweboy The first column looks more like a button to expand the room to see more details about the room members instead of an indicator on whether it is password-protected, that's why I asked. Or is that still under work?

jroweboy commented:
One other issue that probably should be investigated before merge is the search and filter options no longer seem to work. I don't know what broke them, but all the code is still there so it shouldn't be too much to fix it.

jl91569 commented:
I'm still getting multiple error windows opening when trying to connect to a nonexistent server.

1. Connect to a server.
2. Disconnect from the server
3. Fail to connect to a server

Expected behaviour: One error window opens
Actual behaviour: 2 error windows open

Tested on citra-windows-mingw-20180409-e8ac99e

jroweboy commented:
I think that i've fixed everything thats holding this back.

* Lobby Filters are working again.
* The menu items for Show Current Room and Leave Room are functional
* Couldn't replicate @jl91569 issue. Maybe my leave room changes fixed it :P
* Removed the misleading instructions in Direct Connect
* Minor stylistic changes in a few of the UIs

pls do one last review and merge @wwylele

BreadFish64 commented on src/citra_qt/multiplayer/message.cpp:
```diff
@@ -0,0 +1,59 @@
+    QT_TR_NOOP("The host of the room has banned you. Speak with the host to unban you "
+               "or try a different room."));
+const ConnectionError WRONG_VERSION(
+    QT_TR_NOOP("You are using a different version of Citra-Local-Wifi(tm) then the room "
```
*than* the room, the message could be more professional too

BreadFish64 commented on src/citra_qt/multiplayer/message.cpp:
```diff
@@ -0,0 +1,59 @@
+const ConnectionError NO_INTERNET(
+    QT_TR_NOOP("Unable to find an internet connection. Check your internet settings."));
+const ConnectionError UNABLE_TO_CONNECT(
+    QT_TR_NOOP("Unable to connect to the host. Verify that the connection settings are correct."));
```
this usually happens when the host doesn't have the proper port forwarded, maybe mention that?

BreadFish64 commented on src/citra_qt/multiplayer/message.cpp:
```diff
@@ -0,0 +1,59 @@
+const ConnectionError WRONG_VERSION(
+    QT_TR_NOOP("You are using a different version of Citra-Local-Wifi(tm) then the room "
+               "you are trying to connect to."));
+const ConnectionError WRONG_PASSWORD(QT_TR_NOOP("Wrong password."));
```
"Incorrect Password"

BreadFish64 commented on src/citra_qt/multiplayer/state.cpp:
```diff
@@ -0,0 +1,150 @@
+void MultiplayerState::OnAnnounceFailed(const Common::WebResult& result) {
+    announce_multiplayer_session->Stop();
+    QMessageBox::warning(this, tr("Error"),
+                         tr("Failed to announce the room to the public lobby.\nThe room will not "
```
We get quite a few users who don't know what to do about the "credentials needed" error, I'd hate to make a special case for it, but there should be some way to let them know they need to log in with their Citra account or make a private room

wwylele commented on src/citra_qt/multiplayer/validation.h:
```diff
@@ -0,0 +1,28 @@
+static const QRegExp ip_regex(
+    "(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|"
+    "2[0-4][0-9]|25[0-5])");
+static const QValidator* ip = new QRegExpValidator(ip_regex);
```
Are you sure you are going to:
 - `new` a global object and not `delete` it anywhere (or somewhere I can't see) (I know this is harmless in general, but would produce noise if we run citra in memory leak detection tools)
 - and `new` objects for every translation units, one for each, that include this header file?

jroweboy commented on src/citra_qt/game_list_p.h:
```diff
@@ -7,6 +7,7 @@
 #include <atomic>
 #include <map>
 #include <unordered_map>
+#include <QCoreApplication>
```
https://github.com/jroweboy/citra/blob/180568e5dd88d0c0f078f34e493564dd630e26e7/src/citra_qt/game_list_p.h#L161 (i missed adding this header in the quick fix)

jroweboy commented:
Addressed the review comments.

tywald commented:
One issue I encountered on Appveyor build eac98297.
Trying to expand to view players inside the room shows blank:
![image](https://user-images.githubusercontent.com/10337060/38935192-052d1572-431e-11e8-9a9e-1d56fd316586.png)

By opening the Public Room Browser the log says:

jroweboy commented:
Last call for issue reports. If you find any bugs, report them here soon because I plan to have this merged very very soon.

Issues fixed since the last update:
* Host Room and Direct Connect are working again
* Trying to host a room while in a room only gives one warning pop up (and no longer crashes)
* Player list is working again

tywald commented:
Tested with Appveyor build: HEAD-c7fee2c

If you filter anything then remove it, it will break the player list.

BreadFish64 commented on src/citra_qt/multiplayer/message.cpp:
```diff
@@ -0,0 +1,61 @@
+    QT_TR_NOOP("The host of the room has banned you. Speak with the host to unban you "
+               "or try a different room."));
+const ConnectionError WRONG_VERSION(
+    QT_TR_NOOP("Version mismatch! Please update to the latest version of citra. If the problem "
```
capital Citra
